src/decode.jl

```diff
@@ -0,0 +1,147 @@
+  value, e.g., `0.3 => 2/8` and `0.35 => 3/8`.
+
+# Examples
+
+```jldoctest
+julia> using JpegTurbo, TestImages, ImageCore
+
+julia> filename = testimage("earth", download_only=true);
+
+julia> img = jpeg_decode(filename); summary(img)
+"3002×3000 Array{RGB{N0f8},2} with eltype RGB{N0f8}"
+
+julia> img = jpeg_decode(Gray, filename; scale_ratio=0.25); summary(img)
+"751×750 Array{Gray{N0f8},2} with eltype Gray{N0f8}"
+```
+
+For image preview and similar purposes, `T` and `scale_ratio` are useful parameters to
+accelerate the JPEG decoding process. For color JPEG image, `jpeg_decode(Gray, filename)` is
+faster than `jpeg_decode(filename)` since the color components need not be processed.
+Smaller `scale_ratio` permits significantly faster decoding since fewer pixels need be
+processed and a simpler IDCT method can be used.
+
+```julia
+using BenchmarkTools, TestImages, JpegTurbo
+filename = testimage("earth", download_only=true)
+# full decompression
+@btime jpeg_decode(filename); # 224.760 ms (7 allocations: 51.54 MiB)
+# only decompress luminance component
+@btime jpeg_decode(Gray, filename); # 91.157 ms (6 allocations: 17.18 MiB)
+# process only a few pixels
+@btime jpeg_decode(filename; scale_ratio=0.25); # 77.254 ms (8 allocations: 3.23 MiB)
+# process only a few pixels for luminance component
+@btime jpeg_decode(Gray, filename; scale_ratio=0.25); # 63.119 ms (6 allocations: 1.08 MiB)
+```
+
+"""
+function jpeg_decode(
+        ::Type{CT},
+        filename::AbstractString;
+        transpose=false,
+        scale_ratio=1) where CT<:Colorant
+    infile = ccall(:fopen, Libc.FILE, (Cstring, Cstring), filename, "rb")
+    @assert infile.ptr != C_NULL
+    out_CT, jpeg_cls = _jpeg_out_color_space(CT)
+
+    local cinfo, out
+    try
+        cinfo = LibJpeg.jpeg_decompress_struct()
+        cinfo_ptr = Ref(cinfo)
+        jerr = Ref{LibJpeg.jpeg_error_mgr}()
+        cinfo.err = LibJpeg.jpeg_std_error(jerr)
+        LibJpeg.jpeg_create_decompress(cinfo_ptr)
+        LibJpeg.jpeg_stdio_src(cinfo_ptr, infile)
+        LibJpeg.jpeg_read_header(cinfo_ptr, true)
+
+        # set decompression parameters, if given
+        r = _cal_scale_ratio(scale_ratio)
+        cinfo.scale_num, cinfo.scale_denom = r.num, r.den
+        cinfo.out_color_space = jpeg_cls
+
+        # eagerly calculate dimension information so that `output_XXX` fields are valid.
+        LibJpeg.jpeg_calc_output_dimensions(cinfo_ptr)
+        out_size = (Int(cinfo.output_width), Int(cinfo.output_height))
+        if !all(x -> x<=65535, out_size)
+            error("Suspicious inferred image size $out_size: each dimension is expected to have at most 65535 pixels.")
+        end
+        out_ndims = Int(cinfo.output_components)
+        @assert out_ndims == length(out_CT) "Suspicous output color space: $cinfo.out_color_space"
+
+        out = Matrix{out_CT}(undef, out_size)
+        _jpeg_decode!(out, cinfo)
+    finally
+        LibJpeg.jpeg_destroy_decompress(Ref(cinfo))
```

Comment:
```suggestion
        LibJpeg.jpeg_destroy_decompress(cinfo_ptr)
```

Reply:
I use `Ref(cinfo)` here because in this `try ... finally ... end` block, `cinfo_ptr` is not accessible in the `finally` part.

Is `Ref(cinfo)` equivalent to

```julia
local cinfo_ptr
try
    ...
finally
    LibJpeg.jpeg_destroy_decompress(cinfo_ptr)
end
```
?

Reply:
Try to rewrite the snippet in this way:
```
cinfo_ref = Ref{LibJpeg.jpeg_decompress_struct}()
try
         cinfo = cinfo_ref[]
         ...
finally
...
end
```